sources/common/inc/SafeUInt.h

```diff
@@ -0,0 +1,221 @@
+            {
+                throw std::out_of_range("");
+            }
+            return SafeUInt<T>(m_Value - rhs.GetValue());
+        }
+
+        /// Overrides the substraction operator '-='.
+        SafeUInt<T> &operator-=(const SafeUInt<T> &rhs)
+        {
+            if (m_Value < rhs.GetValue())
+            {
+                throw std::out_of_range("");
+            }
+            m_Value -= rhs.GetValue();
+            return *this;
+        }
+
+        /// Overrides the multiplication operator '*'.
+        SafeUInt<T> operator*(uint64_t rhs) const
+        {
+            if (m_Value)
+            {
+                if ((std::numeric_limits<T>::max() / m_Value) < rhs)
+                {
+                    throw std::out_of_range("");
+                }
+            }
+            return SafeUInt<T>(m_Value * rhs);
+        }
+
+        /// Overrides the multiplication operator '*= uint64_t'.
+        SafeUInt<T> &operator*=(uint64_t rhs)
+        {
+            if (m_Value)
+            {
+                if ((std::numeric_limits<T>::max() / m_Value) < rhs)
+                {
+                    throw std::out_of_range("");
+                }
+            }
+            m_Value *= rhs;
+            return *this;
+        }
+
+        /// Overrides the multiplication operator '* double'.
+        SafeUInt<T> operator*(double rhs) const
+        {
+            if (m_Value)
+            {
+                if ((std::numeric_limits<T>::max() / m_Value) < rhs)
+                {
+                    throw std::out_of_range("");
+                }
+            }
+            return SafeUInt<T>(static_cast<T>(m_Value * rhs));
+        }
+
+        /// Overrides the multiplication operator '*= double'.
+        SafeUInt<T> &operator*=(double rhs)
+        {
+            if (m_Value)
+            {
+                if ((std::numeric_limits<T>::max() / m_Value) < rhs)
+                {
+                    throw std::out_of_range("");
+                }
+            }
+            m_Value *= rhs;
+            return *this;
+        }
+
+        /// Overrides the division operator '/'.
+        SafeUInt<T> operator/(uint64_t rhs) const
+        {
+            return SafeUInt<T>(m_Value / rhs);
+        }
+
+        /// Overrides the division operator '=/'.
+        SafeUInt<T> &operator/=(uint64_t rhs)
+        {
+            m_Value /= rhs;
+            return *this;
+        }
+
+        /// Returns the maximum type value.
+        static SafeUInt<T> max()
+        {
+            return SafeUInt<T>(std::numeric_limits<T>::max());
+        }
+
+        /// Overrides the output stream operator.
+        template<T> friend std::ostream &operator<<(std::ostream &out, SafeUInt<T> &ts);
+
+        /// Overrides the multiplication operator.
+        template<T> friend SafeUInt<T> operator*(uint64_t lhs, const SafeUInt<T> &rhs);
+
+    private:
+
+        /// The unsigned integer value.
+        T m_Value;
+};
```

Comment:
I'm sorry but I don't like this class. It not generic and contains a lot of duplicated code.